### ISSUE TITLE
Reason for multiple initialisation should not be called simultaneously?

### ISSUE THREAD

**Issue by jihlee** [MvvmCross/MvvmCross#955]
In MvxAndroidSetupSingleton, EnsureInitialized() is designed not to be called simultaneously.
In detail, if one thread is running _setup.Initialize() and other threads call EnsureInitialized() again, it goes into locked critical section, and throw MvxException("Multiple initialize calls made") since initialisation is not ended.

My application has GCM notification service and it updates the app's view. It was developed by IMvxMessenger, and I should call EnsureInitialized() of MvxAndroidSetupSingleton in GCMIntentService.

The thing is that if the application is starting and a GCM message is received at that moment, EnsureInitialized() is called again by GCM Service, and MvxException is thrown.

To resolve the issue, I would like to change EnsureInitilized() work as
- It can be called simultaneously
- Mvx system is initialised only once though.
- If the method is called multiple times, the first caller does initialise the system and others just wait there.
- The first caller notify others when the system is prepared.

Thanks

**Comment by Cheesebaron** [comment]
Reopen this with more information.

**Comment by monesprit54** [comment]
Added description of the issue and reopen the issue.
I submitted first ugly empty issue by mistake :( 

Thanks for fast reply!!

**Comment by monesprit54** [comment]
Please confirm the suggestion would be feasible or not. Is there any reason the method should not be called multiple times?

**Comment by slodge** [comment]
I'm not sure I understand how you intend to implement this.

I'm especially worried about the line "others just wait there" - not sure how they are waiting - is this async awaiting? is it UI thread blocking? something else? Blocking threads often leads to other problems :(

Can you demonstrate what you propose - e.g. it should be possible to write a small wrapper class (your own "setup singleton") which should have the behaviour you are looking for?

Alternatively, it might be possible to throw a more specific error? Or it might be possible to expose an event (or for you to hook into an existing setup event)? Together these things would allow you to implement your own mechanism for pausing and then processing the setup queue.

**Comment by davidschwegler** [comment]
https://github.com/MvvmCross/MvvmCross/blob/3.5/Cirrious/Cirrious.MvvmCross.Droid/Platform/MvxAndroidSetupSingleton.cs#L36

I'm actually in the same boat as monespirit54. 

When the user starts the app via the main launcher/splash screen, it does an async load of MvvmCross. Any calls to EnsureInitialized will cause an exception during this time. 

However, if a GCM message comes in at that exact moment, we won't know that the environment is initializing, and will call EnsureIntialized (in order to load the app and process the GCM message), which will crash.

@slodge  Is there any reason why EnsureInitialized shouldn't just block until the async initialization is complete (it would currently crash, so blocking is better in this rare scenario)? EnsureInitialized is also called by MvxActivityViewExtensions.OnViewCreate, which seems like it would be very difficult to transition to waiting async for completion (the scenario where that would run into this problem if the user started the app's splash screen, then click something like a notification to launch a new part of the app). TLDR: IMHO, block > crash

I could submit a pull request with the above, but it's not high enough priority for me currently. I'm currently just crossing my fingers that it is an extremely rare situation and will "never" happen. :) @monesprit54 if this is a frequent issue for you, you might have to just build the MvvmCross libraries and customize that yourself.

**Comment by slodge** [comment]
> is there any reason why EnsureInitialized shouldn't just block until the async initialization is complete 

I'm pretty sure if we change the code so that the UI thread blocks then we'll find that setup doesn't complete - because setup needs the UI thread for some operations. (But I could be wrong - v v v happy for people to experiment :)).

If the issue is with "how to get GCM message receivers to wait until after setup has completed", then I'd guess some pattern could be used within the GCM message receiver to handle the exception (or change it to a return value) and to then find (or create) an async hook to allow the receiver to wait for the setup to complete.

A good first step might be to find some way to get a reproducible and shareable test case. Not sure quite how though... maybe we could try some sort of local notification test app to try to repro? Open to ideas

**Comment by monesprit54** [comment]
Sadly, this issue gives me a lot of crashes. Like 10 crashes per day for production app at the moment.
I will first try to customize the setup file as @slodge suggested, and go forward  to shareable test case.

**Comment by davidschwegler** [comment]
You can easily simulate the scenario simply by posting a call to EnsureInitialized on the UI thread from within a backgrounded Setup method. 

See https://github.com/MvvmCross/MvvmCross/blob/3.5/Cirrious/Cirrious.MvvmCross.Droid/Platform/MvxAndroidSetupSingleton.cs#L89

My GCM message processing class is static, so I can just call it with a fake message to more exactly simulate the scenario, but that's irrelevant. For example:



```
Setup.cs
...
public override voide InitializeSecondary()
{
Application.SynchronizationContext.Post(ignored => 
  // this could be a call to your gcm message processing code where you call EnsureInitialized
  MvxAndroidSetupSingleton.EnsureSingletonAvailable(Application.Context).EnsureInitialized()), null);

 base.InitializeSecondary();
}
```

Then just launch the app and watch it crash.

**Comment by davidschwegler** [comment]
@monesprit54  I just added some code to workaround this crash in our app. I think the general pattern could be adapted to the MvvmCross code in a ```public Task EnsureInitializedAsync()``` in MvxAndroidSetupSingleton (removing the exception handling, of course, and hooking into the real initialization methods). This code assumes callers will be on the UI thread (GCM/intents always will be), but I think you could add some locks to support async calls if needed.

```
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Android.App;
using Android.Util;
using Cirrious.CrossCore.Exceptions;
using Cirrious.MvvmCross.Droid.Platform;

namespace OurNamespace.Utility
{
	public static class MvxAndroidSetupSingletonUtility
	{
		/// <summary>
		/// Initializes the MvvmCross system, avoiding the "Multiple initialize calls made" crash.
		/// Must be called from the UI thread.
		/// If the app is already initializing async (via the splash screen), attempting to call MvxAndroidSetupSingleton.EnsureInitialized would normally crash
		/// https://github.com/MvvmCross/MvvmCross/issues/955
		/// https://github.com/MvvmCross/MvvmCross/blob/3.5/Cirrious/Cirrious.MvvmCross.Droid/Platform/MvxAndroidSetupSingleton.cs#L39
		/// https://github.com/MvvmCross/MvvmCross/blob/3.5/Cirrious/Cirrious.MvvmCross.Droid/Platform/MvxAndroidSetupSingleton.cs#L89
		/// </summary>
		public static async Task SafeEnsureInitialized()
		{
			if (s_awaitingInitializers != null)
			{
				var completionSource = new TaskCompletionSource<bool>();
				s_awaitingInitializers.Add(completionSource);

				Log.Debug(Tag, "SafeEnsureInitialized: Adding waiter, waiting count = " + s_awaitingInitializers.Count);
				await completionSource.Task;

				return;
			}

			Log.Debug(Tag, "SafeEnsureInitialized: No current waiters");
			s_awaitingInitializers = new List<TaskCompletionSource<bool>>();
			while (true)
			{
				try
				{
					Log.Debug(Tag, "SafeEnsureInitialized: Trying EnsureInitialized()");
					MvxAndroidSetupSingleton.EnsureSingletonAvailable(Application.Context).EnsureInitialized();

					if (s_awaitingInitializers != null)
					{
						var copy = s_awaitingInitializers;
						s_awaitingInitializers = null;

						// post this so the current call completes before the others
						Log.Debug(Tag, "SafeEnsureInitialized: Success, posting completion result to " + copy.Count + " waiters");
						Application.SynchronizationContext.Post(ignored =>
						{
							foreach (var completionSource in copy)
								completionSource.SetResult(true);
						}, null);
					}

					break;
				}
				catch (MvxException e)
				{
					Log.Debug(Tag, "Caught MvxException, assuming it is 'Multiple initialize calls made': " + e);
				}

				await Task.Delay(TimeSpan.FromSeconds(1));
			}
		}

		static readonly string Tag = typeof(MvxAndroidSetupSingletonUtility).Name;

		static List<TaskCompletionSource<bool>> s_awaitingInitializers;
	}
}
```

**Comment by monesprit54** [comment]
Thank you @davidschwegler ! I think it is well written and good to go.

I think this kind of solution is feasible solution if we do not change the MvxAndroidSetupSingleton code since some plugins have dependency over it. They could be rewritten in user application projects, but I think that it is not the final solution for this issue.

**Comment by andyci** [comment]
@davidschwegler please can you create a pull request for that helper.  I've used it in my project and it works great.  Without this it doesn't seem possible to use a sticky Android Service as if you call EnsureInitialized() it will throw an exception in certain scenarios and if you don't call it then IOC, etc. won't be initialized properly.

**Comment by davidschwegler** [comment]
@andyci MvvmCross' initialization code needs to be modified to make it thread-safe - the goal of MvvmCross' API *should* be that calling MvxAndroidSetupSingleton.EnsureSingletonAvailable is enough in all cases. This hack I wrote is not a good API or a good solution (guessing about MvxExceptions and trying again after timeouts is not is a correct way to solve the issue), so I don't think it should be in MvvmCross itself. 

I'd suggest just copying it for use as a helper class until someone fixes MvvmCross itself.

**Comment by sbokatuk** [comment]
I can reproduce issue on mvvmcross 5.6.3,
sample is here https://github.com/sbokatuk/mvvmcross955/

to reproduce crash you neeed to

1) install app, reboot, wait till recieve notification in notification bar - this means service is running
2) start app, if not crashed goto 3
3) kill app(close app from tasks manager) and faster click on app icon to run - 80% it will crash

But! if we modify this row from 'void' [here](https://github.com/sbokatuk/mvvmcross955/blob/master/CrashableTestApp.Droid/Services/TestTimerService.cs#L37) to 'async Task' [sample](https://github.com/sbokatuk/mvvmcross955/blob/nocrash/CrashableTestApp.Droid/Services/TestTimerService.cs#L37) app will stop crashing

`[Service]
    public class TestTimerService : Service
    {
       public override void OnCreate()
        {
            TestStuff();
        }

        //replace void to async Task and you will see a magic
        private void TestStuff()
        {
            Console.WriteLine($"Thread: {System.Threading.Thread.CurrentThread.ManagedThreadId}");
            AndroidMvxInitializer.SafeEnsureInitialized();
        }
    }`

`public static class AndroidMvxInitializer
    {
        /// <summary>
        /// If the app is already initializing async (via the splash screen), attempting to call MvxAndroidSetupSingleton.EnsureInitialized would normally crash
        /// https://github.com/MvvmCross/MvvmCross/issues/955
        /// </summary>
        public static void SafeEnsureInitialized([CallerMemberName]string caller = null)
        {
            System.Console.WriteLine("{0} SafeEnsureInitialized: Trying EnsureInitialized() from {1}", nameof(AndroidMvxInitializer), caller);
            MvxAndroidSetupSingleton.EnsureSingletonAvailable(Application.Context).EnsureInitialized();
        }
    }`

**Comment by nmilcoff** [comment]
@sbokatuk Plase open a new issue linking this one there. This is already closed.
